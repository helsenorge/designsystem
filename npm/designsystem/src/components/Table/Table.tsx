import React from 'react';
import classNames from 'classnames';

import tableStyles from './styles.module.scss';

import { useIsVisible } from '../../hooks/useIsVisible';

export enum SmallViewportVariant {
  horizontalscroll = 'horizontalscroll',
  block = 'block',
}

export interface Props {
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the content of the table. Use TableHaed and TableBody */
  children: React.ReactNode;
  /** Unique ID */
  id?: string;
  /** Id used for testing */
  testId?: string;
  /** Type of table view om small devices */
  smallViewportVariant?: SmallViewportVariant;
}

export const Table = function Table({ id, testId, className, children, smallViewportVariant = SmallViewportVariant.block }: Props) {
  const tableClass = classNames(
    tableStyles['table'],
    { [tableStyles['table--collapse2col']]: smallViewportVariant === SmallViewportVariant.block },
    className
  );

  return (
    <table className={tableClass} id={id} data-testid={testId}>
      {children}
    </table>
  );
};

export default Table;
