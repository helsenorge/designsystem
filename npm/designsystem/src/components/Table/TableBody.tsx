import React from 'react';

interface Props {
  children: React.ReactNode;
}

export const TableBody = function TableBody({ children }: Props) {
  return <tbody>{children}</tbody>;
};

export default TableBody;
