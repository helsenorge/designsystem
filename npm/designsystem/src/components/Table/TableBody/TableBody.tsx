import React from 'react';

import classNames from 'classnames';

import tableStyles from '../styles.module.scss';
import { ModeType } from '../Table';

export interface Props {
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the content of the table body. Use TableRows */
  children?: React.ReactNode;
  /** For display with less space. Discouraged to use together with interactive elements. */
  mode?: ModeType;
}

export const TableBody = ({ className, children, mode = ModeType.normal }: Props): React.JSX.Element => {
  const tableBodyClasses = classNames(tableStyles['table-body'], className);
  return (
    <tbody className={tableBodyClasses}>
      {React.Children.map(children, child => React.isValidElement<{ mode?: ModeType }>(child) && React.cloneElement(child, { mode }))}
    </tbody>
  );
};

export default TableBody;
