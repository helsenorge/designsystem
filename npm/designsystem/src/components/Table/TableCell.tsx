import React from 'react';
import classNames from 'classnames';
import tableStyles from './styles.module.scss';

interface Props {
  nowrap?: boolean;
  dataLabel: string;
  textAlign?: 'left' | 'center' | 'right';
  /** Adds custom classes to the element. */
  className?: string;
  children: React.ReactNode;
}

export const TableCell = function TableCell({ nowrap = false, textAlign = 'left', dataLabel, children, className }: Props) {
  const tableCellClass = classNames(
    tableStyles['table__cell'],
    { [tableStyles['table__cell--nowrap']]: nowrap },
    { [tableStyles['table__cell--center']]: textAlign === 'center' },
    { [tableStyles['table__cell--right']]: textAlign === 'right' },
    className
  );

  return (
    <td className={tableCellClass} data-label={dataLabel}>
      {children}
    </td>
  );
};

export default TableCell;
