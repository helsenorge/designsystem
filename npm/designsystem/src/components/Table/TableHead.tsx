import React from 'react';
import classNames from 'classnames';
import tableStyles from './styles.module.scss';

interface Props {
  sortable?: boolean;
  category?: 'normal' | 'transparent';
  /** Adds custom classes to the element. */
  className?: string;
  children: React.ReactNode;
}

export const TableHead = function TableHead({ sortable = false, category = 'normal', className, children }: Props) {
  const tableHeadClass = classNames(
    tableStyles['table__head'],
    {
      [tableStyles['table__head--normal']]: category === 'normal',
      [tableStyles['table__head--transparent']]: category === 'transparent',
      [tableStyles['table__head--sortable']]: sortable,
    },
    className
  );

  return <thead className={tableHeadClass}>{children}</thead>;
};

export default TableHead;
