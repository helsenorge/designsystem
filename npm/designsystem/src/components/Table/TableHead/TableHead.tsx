import React from 'react';

import classNames from 'classnames';

import tableStyles from '../styles.module.scss';
import { ModeType } from '../Table';

export enum HeaderCategory {
  normal = 'normal',
  transparent = 'transparent',
  sortable = 'sortable',
}

export interface Props {
  /** Header category for styling */
  category: HeaderCategory;
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the content of the thead. Add table rows  */
  children: React.ReactNode;
  /** For display with less space. Discouraged to use together with interactive elements. */
  mode?: ModeType;
}

export const TableHead = ({ category, className, children, mode }: Props): React.JSX.Element => {
  const tableHeadClass = classNames(
    tableStyles['table__head'],
    {
      [tableStyles['table__head--normal']]: category === HeaderCategory.normal,
      [tableStyles['table__head--transparent']]: category === HeaderCategory.transparent,
      [tableStyles['table__head--sortable']]: category == HeaderCategory.sortable,
      [tableStyles['table__head--compact']]: mode == ModeType.compact,
    },
    className
  );

  return (
    <thead className={tableHeadClass}>
      {React.Children.map(children, child => React.isValidElement<Props>(child) && React.cloneElement(child, { mode }))}
    </thead>
  );
};

export default TableHead;
