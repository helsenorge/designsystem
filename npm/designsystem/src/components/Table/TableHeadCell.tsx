import React from 'react';
import classNames from 'classnames';
import tableStyles from './styles.module.scss';
import Icon, { IconSize } from '../Icons';
import ArrowUp from '../Icons/ArrowTop';
import ArrowDown from '../Icons/ArrowBottom';

interface Props {
  sortable?: boolean;
  transparent?: boolean;
  sortDir?: 'asc' | 'desc';
  onClick?: () => void;
  /** Adds custom classes to the element. */
  className?: string;
  children?: React.ReactNode;
}

export const TableHeadCell = function TableHeadCell({ sortable, onClick, className, children, sortDir }: Props) {
  const tableHeadCellDefaultClass = tableStyles['table__head-cell'];

  if (!sortable) {
    return (
      <th scope="col" className={classNames(tableHeadCellDefaultClass, className)}>
        {children}
      </th>
    );
  }

  const tableHeadCellClass = classNames(
    tableHeadCellDefaultClass,
    tableStyles['table__head-cell--sortable'],
    { [tableStyles['table__head-cell--sorted']]: !!sortDir },
    className
  );

  let sortIcon: JSX.Element | null = null;
  if (sortDir) {
    const arrow = sortDir == 'asc' ? ArrowDown : ArrowUp;
    sortIcon = (
      <div>
        <Icon svgIcon={arrow} size={IconSize.XXSmall}></Icon>
      </div>
    );
  }

  return (
    <th
      scope="col"
      role="button"
      tabIndex={1}
      className={tableHeadCellClass}
      onClick={onClick}
      onKeyPress={e => e.key === 'Enter' && onClick && onClick()}
    >
      {sortIcon}
      {children}
    </th>
  );
};

export default TableHeadCell;
