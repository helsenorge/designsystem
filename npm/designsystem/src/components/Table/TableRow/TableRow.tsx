import React from 'react';

import classNames from 'classnames';

import tableStyles from '../styles.module.scss';
import { ModeType } from '../Table';
import TableExpanderCellMobile from '../TableExpanderCell/TableExpanderCellMobile';

export interface Props {
  /** Sets if expanded row can be expanded */
  expandable?: boolean;
  /** Sets if expanded row is expanded */
  expanded?: boolean;
  /** When hide/show button for expanded row is clicked. */
  onClick?: () => void;
  /** Text for expanded row hide button. */
  hideDetailsText?: string;
  /** Text for expanded row show button. */
  showDetailsText?: string;
  /** Key attribute for row */
  rowKey?: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the cells of the table row element.  */
  children?: React.ReactNode;
  /** For display with less space. Discouraged to use together with interactive elements. */
  mode?: ModeType;
}

export const TableRow: React.FC<Props> = ({
  rowKey,
  hideDetailsText,
  showDetailsText,
  expandable,
  expanded,
  onClick,
  className,
  children,
  mode = ModeType.normal,
}) => {
  const tableRowClass = classNames(
    tableStyles['table-row'],
    {
      [tableStyles['table__row--expandable']]: expandable,
      [tableStyles['table__row--expanded']]: expanded,
    },
    className
  );

  return (
    <tr className={tableRowClass} onClick={onClick} key={rowKey}>
      {React.Children.map(children, child => React.isValidElement<Props>(child) && React.cloneElement(child, { mode }))}
      {expandable && (
        <TableExpanderCellMobile
          expanded={expanded}
          onClick={onClick}
          hideDetailsText={hideDetailsText}
          showDetailsText={showDetailsText}
          mode={mode}
        />
      )}
    </tr>
  );
};

export default TableRow;
