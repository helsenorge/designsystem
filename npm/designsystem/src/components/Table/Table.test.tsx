import React from 'react';
import { screen, render } from '@testing-library/react';
import { SmallViewportVariant, Table } from './Table';
import { TableBody } from './TableBody';
import { TableCell } from './TableCell';
import { HeaderCategory, TableHead } from './TableHead';
import { TableHeadCell } from './TableHeadCell';
import { TableRow } from './TableRow';

describe('Gitt at Tabel skal vises', (): void => {
  describe('Når den skal vises', (): void => {
    it('Så vises en tabell', (): void => {
      const { container } = render(
        <Table smallViewportVariant={SmallViewportVariant.horizontalscroll} testId="test01">
          <TableHead category={HeaderCategory.normal}>
            <TableRow key="0">
              <TableHeadCell>Navn</TableHeadCell>
              <TableHeadCell>Beskrivelse</TableHeadCell>
            </TableRow>
          </TableHead>
          <TableBody>
            <TableRow key="1">
              <TableCell dataLabel="Navn">Hans Nilsen</TableCell>
              <TableCell dataLabel="Beskrivelse">En ganske lang beskrivelse...</TableCell>
            </TableRow>
          </TableBody>
        </Table>
      );

      const table = screen.getByTestId('test01');
      const child = screen.getByText('Hans Nilsen');

      expect(table).toHaveClass('table');
      expect(child).toHaveClass('table__cell');

      expect(container).toMatchSnapshot();
    });
  });
});
