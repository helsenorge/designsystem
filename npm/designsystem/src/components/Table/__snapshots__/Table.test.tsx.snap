// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at Tabel skal vises Når den skal vises Så vises en tabell 1`] = `
<div>
  <table
    class="table"
    data-testid="test01"
  >
    <thead
      class="table__head table__head--normal"
    >
      <tr
        class="table-row"
      >
        <th
          class="table__head-cell"
          scope="col"
        >
          Navn
        </th>
        <th
          class="table__head-cell"
          scope="col"
        >
          Beskrivelse
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="table-row"
      >
        <td
          class="table__cell"
          data-label="Navn"
        >
          Hans Nilsen
        </td>
        <td
          class="table__cell"
          data-label="Beskrivelse"
        >
          En ganske lang beskrivelse...
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
